### PR TITLE
Add new functions to ParallelSTL

### PR DESCRIPTION
**Pull request by scober**
Specifically: copy, copy_if, and transform. I have been wishing I had these for a while.

I am not super familiar with templated code or the galois parallel interfaces so it is likely that I didn't write these with best practices.

It is also possible that @ddn0 is not the right person to review this, but he was recommended because he was the last person to edit a lot of this file during his re-formatting. @ddn0, feel free to tell me someone else should be reviewing this.

### CODE REVIEW

File: libgalois/include/katana/ParallelSTL.h [hunk 1]
```diff
@@ -391,6 +393,103 @@ partial_sum(InputIt first, InputIt last, OutputIt d_first) {
...
+  static_assert(
+      std::is_base_of_v<std::random_access_iterator_tag, output_category>);
+
+  using diff_type = typename std::iterator_traits<InputIt>::difference_type;
+
+  diff_type length = std::distance(first, last);
+  katana::do_all(
+      katana::iterate(diff_type{0}, length),
+      [&](diff_type index) { *(d_first + index) = unary_op(*(first + index)); },
+      katana::steal());
```
Comment (roshandathathri) [inline libgalois/include/katana/ParallelSTL.h]: Don't use `steal()` for _operators_ that are not imbalanced. Typically, imbalanced operators iterate over edges (which could be imbalanced).

File: libgalois/include/katana/ParallelSTL.h [hunk 2]
```diff
@@ -391,6 +393,103 @@ partial_sum(InputIt first, InputIt last, OutputIt d_first) {
...
+  static_assert(
+      std::is_base_of_v<std::random_access_iterator_tag, output_category>);
+
+  using diff_type = typename std::iterator_traits<InputIt>::difference_type;
+
+  diff_type length = std::distance(first, last);
+  katana::do_all(
+      katana::iterate(diff_type{0}, length),
+      [&](diff_type index) { *(d_first + index) = *(first + index); },
+      katana::steal());
```
Comment (roshandathathri) [inline libgalois/include/katana/ParallelSTL.h]: Ditto

File: libgalois/include/katana/ParallelSTL.h [hunk 3]
```diff
@@ -391,6 +393,103 @@ partial_sum(InputIt first, InputIt last, OutputIt d_first) {
...
+      std::is_base_of_v<std::random_access_iterator_tag, input_category>);
+  static_assert(
+      std::is_base_of_v<std::random_access_iterator_tag, output_category>);
+
+  using diff_type = typename std::iterator_traits<InputIt>::difference_type;
+
+  diff_type length = std::distance(first, last);
+  katana::do_all(
+      katana::iterate(diff_type{0}, length),
+      [&](diff_type index) { *(d_first + index) = *(first + index); },
```
Comment (roshandathathri) [inline libgalois/include/katana/ParallelSTL.h]: Particularly for this but also for the other functions here, I hope this is not precluding vectorization. We may need to revisit this later.

File: libgalois/include/katana/ParallelSTL.h [hunk 4]
```diff
@@ -391,6 +393,103 @@ partial_sum(InputIt first, InputIt last, OutputIt d_first) {
...
+  });
+
+  // second on_each, do copy
+  katana::on_each([&](unsigned tid, unsigned total) {
+    auto [begin, end] = katana::block_range(diff_type{0}, length, tid, total);
+    diff_type count = 0;
+    InputIt it = first + begin;
+    InputIt finish = first + end;
+    diff_type thread_offset = prefix_sum[tid];
+    diff_type num_handled = 0;
```
Comment (roshandathathri) [inline libgalois/include/katana/ParallelSTL.h]: Initialize to `thread_offset` to simplify code below

File: libgalois/include/katana/ParallelSTL.h [hunk 5]
```diff
@@ -391,6 +393,103 @@ partial_sum(InputIt first, InputIt last, OutputIt d_first) {
...
+        count++;
+      }
+    }
+    prefix_sum[tid] = count;
+  });
+
+  // second on_each, do copy
+  katana::on_each([&](unsigned tid, unsigned total) {
+    auto [begin, end] = katana::block_range(diff_type{0}, length, tid, total);
+    diff_type count = 0;
```
Comment (roshandathathri) [inline libgalois/include/katana/ParallelSTL.h]: Unused

File: libgalois/include/katana/ParallelSTL.h [hunk 6]
```diff
@@ -391,6 +393,103 @@ partial_sum(InputIt first, InputIt last, OutputIt d_first) {
...
+    InputIt it = first + begin;
+    InputIt finish = first + end;
```
Comment (roshandathathri) [inline libgalois/include/katana/ParallelSTL.h]:
```cpp
begin += first;
end += first;
```
Do the same for the inspection loop above as well.

**Review by roshandathathri — APPROVED** [review]
LGTM

File: libgalois/include/katana/ParallelSTL.h [hunk 7]
```diff
@@ -391,6 +393,103 @@ partial_sum(InputIt first, InputIt last, OutputIt d_first) {
...
+  using diff_type = typename std::iterator_traits<InputIt>::difference_type;
+
+  // first on_each, set ranges
+  std::vector<diff_type> prefix_sum(getActiveThreads());
+  diff_type length = std::distance(first, last);
+  katana::on_each([&](unsigned tid, unsigned total) {
+    auto [begin, end] = katana::block_range(diff_type{0}, length, tid, total);
+    diff_type count = 0;
+    for (diff_type i = begin; i < end; ++i) {
+      if (pred(*(first + i))) {
```
Comment (ddn0) [inline libgalois/include/katana/ParallelSTL.h]: too cool for `first[i]`?

File: libgalois/include/katana/ParallelSTL.h [hunk 8]
```diff
@@ -391,6 +393,103 @@ partial_sum(InputIt first, InputIt last, OutputIt d_first) {
...
+  static_assert(
+      std::is_base_of_v<std::random_access_iterator_tag, output_category>);
+
+  using diff_type = typename std::iterator_traits<InputIt>::difference_type;
+
+  // first on_each, set ranges
+  std::vector<diff_type> prefix_sum(getActiveThreads());
+  diff_type length = std::distance(first, last);
+  katana::on_each([&](unsigned tid, unsigned total) {
+    auto [begin, end] = katana::block_range(diff_type{0}, length, tid, total);
```
Comment (ddn0) [inline libgalois/include/katana/ParallelSTL.h]: IIRC, `katana::block_range` should work on any random access iterators: `katana::block_range(first, last, tid, total)`

File: libgalois/include/katana/ParallelSTL.h [hunk 9]
```diff
@@ -391,6 +393,103 @@ partial_sum(InputIt first, InputIt last, OutputIt d_first) {
...
+    diff_type count = 0;
+    for (diff_type i = begin; i < end; ++i) {
+      if (pred(*(first + i))) {
+        count++;
+      }
+    }
+    prefix_sum[tid] = count;
```
Comment (ddn0) [inline libgalois/include/katana/ParallelSTL.h:474]: Why is this called `prefix_sum`? This just seems like counts per thread.

For the code below to work wouldn't we need to add a loop like:
```cpp
for (size_t i = 1, n = prefix_sum.size(); i < n; ++i) {
  prefix_sum[i] += prefix_sum[i-1];
}
```
?

File: libgalois/include/katana/ParallelSTL.h (hunk 3 above)
Comment (scober) [reply libgalois/include/katana/ParallelSTL.h]: We could just do a series of `std::copy`s in an `on_each`. Then we would get the benefits of the stl's optmizations.

Comment (roshandathathri) [reply libgalois/include/katana/ParallelSTL.h]: That's a good idea :)

File: libgalois/include/katana/ParallelSTL.h (hunk 7 above)
Comment (scober) [reply libgalois/include/katana/ParallelSTL.h]: I only write "real C"

File: libgalois/include/katana/ParallelSTL.h (hunk 3 above)
Comment (scober) [reply libgalois/include/katana/ParallelSTL.h]: I went ahead and made this change

File: libgalois/include/katana/ParallelSTL.h [hunk 10]
```diff
@@ -391,6 +393,102 @@ partial_sum(InputIt first, InputIt last, OutputIt d_first) {
   }
 }
 
+template <class InputIt, class OutputIt, class UnaryOperation>
+OutputIt
+transform(
+    InputIt first, InputIt last, OutputIt d_first, UnaryOperation unary_op) {
+  // This will only work for random access iterators
```
Comment (ddn0) [inline libgalois/include/katana/ParallelSTL.h]: FWIW, this comment seems a little redundant wrt to the static_assert below

If you want, you could write just:
```cpp
static_assert(...., "transform is only supported on random access iterators");
```

(and throughout)

File: libgalois/include/katana/ParallelSTL.h [hunk 11]
```diff
@@ -391,6 +393,102 @@ partial_sum(InputIt first, InputIt last, OutputIt d_first) {
...
+    for (; first != last; ++first) {
+      if (pred(first)) {
+        count++;
+      }
+    }
+    prefix_sum[tid] = count;
+  });
+
+  // calculate prefix sums
+  for (size_t tid = 1; tid < prefix_sum.size(); ++tid) {
```
Comment (ddn0) [inline libgalois/include/katana/ParallelSTL.h]: BTW, https://en.cppreference.com/w/cpp/algorithm/partial_sum

(an explicit loop is fine as well)

small preference for avoiding `prefix_sum.size()` in the loop condition though: `for (size_t tid = ...., n = ...; tid < n; ++tid) ....`

File: libgalois/include/katana/ParallelSTL.h [hunk 12]
```diff
@@ -391,6 +393,102 @@ partial_sum(InputIt first, InputIt last, OutputIt d_first) {
...
+  katana::on_each([&](unsigned tid, unsigned total) {
+    auto [begin, end] = katana::block_range(first, last, tid, total);
+
+    diff_type offset = tid == 0 ? 0 : prefix_sum[tid - 1];
+    OutputIt actual_end = std::copy_if(begin, end, d_first + offset, pred);
+
+    KATANA_LOG_DEBUG_ASSERT(actual_end == prefix_sum[tid]);
+  });
+
+  return d_first + prefix_sum[num_threads - 1];
```
Comment (ddn0) [inline libgalois/include/katana/ParallelSTL.h]: also fine as is, but `prefix_sum[num_threads - 1]` = `prefix_sum.back()`

**Review by ddn0 — APPROVED** [review]
Gucci

File: libgalois/include/katana/ParallelSTL.h [hunk 13]
```diff
@@ -391,6 +393,103 @@ partial_sum(InputIt first, InputIt last, OutputIt d_first) {
...
+      }
+    }
+    prefix_sum[tid] = count;
+  });
+
+  // calculate prefix sums
+  std::partial_sum(prefix_sum.begin(), prefix_sum.end(), prefix_sum.begin());
+
+  // second on_each, do copy_if
+  katana::on_each([&](unsigned tid, unsigned total) {
```
Comment (roshandathathri) [inline libgalois/include/katana/ParallelSTL.h]: Drop `katana::` to be consistent

**Pull request merged**